<div style="border:solid green 2px; padding: 20px">
<b>Привет, Софья!</b>

Меня зовут Александр, и я буду проверять твой проект. Предлагаю общаться на «ты» :) Но если это не удобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать data science. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - **пожалуйста, не перемещай, не изменяй и не удаляй их**. Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самой найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Так выделены небольшие замечания. Я надеюсь, что их ты тоже учтешь - твой проект от этого станет только лучше. Но настаивать на их отработке не буду.

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом.** Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#LogisticRegression" data-toc-modified-id="LogisticRegression-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>LogisticRegression</a></span></li><li><span><a href="#DecisionTreeClassifier" data-toc-modified-id="DecisionTreeClassifier-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>DecisionTreeClassifier</a></span></li><li><span><a href="#LGBMClassifier" data-toc-modified-id="LGBMClassifier-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>LGBMClassifier</a></span></li><li><span><a href="#CatBoostClassifier---не-вытягивает-ядро-тренажера,-каждый-раз-при-запуске-пишет,-&quot;Ядро-умерло&quot;" data-toc-modified-id="CatBoostClassifier---не-вытягивает-ядро-тренажера,-каждый-раз-при-запуске-пишет,-&quot;Ядро-умерло&quot;-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>CatBoostClassifier - не вытягивает ядро тренажера, каждый раз при запуске пишет, "Ядро умерло"</a></span></li></ul></li><li><span><a href="#Тестирование" data-toc-modified-id="Тестирование-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Тестирование</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

In [1]:
import pandas as pd
import numpy as np
import nltk
from nltk.stem import WordNetLemmatizer 
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.corpus import stopwords
from sklearn.linear_model import LogisticRegression
from catboost import CatBoostClassifier
from lightgbm import LGBMClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, confusion_matrix
import warnings
from sklearn.linear_model import LinearRegression 
from sklearn.tree import DecisionTreeClassifier
warnings.filterwarnings('ignore')
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
nltk.download('wordnet')
nltk.download('stopwords')
nltk.download('punkt')
import re
from nltk.corpus import stopwords as nltk_stopwords

[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /home/jovyan/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [2]:
data = pd.read_csv('/datasets/toxic_comments.csv')
data.head(5)

Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Огонь, данные на месте;)</div>

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


Проверим баланс классов

In [4]:
data['toxic'].value_counts()

0    143106
1     16186
Name: toxic, dtype: int64

**Вывод**
Классы несбалансированы. Необходимо учитывать это при обучении моделей.

Напишем функции для очистки и леммитизации 

In [5]:
def clear_text(text):
    re_list = re.sub(r"[^a-zA-Z']", ' ', text)
    re_list = re_list.split()
    re_list = " ".join(re_list)
    return re_list

m = WordNetLemmatizer()

def lemmatize_text(text):
    word_list = nltk.word_tokenize(text)
    
    return ' '.join([m.lemmatize(w) for w in word_list])

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


In [7]:
data['text'] = data['text'].apply(clear_text)
data['lemm_text'] = data['text'].apply(lemmatize_text)
data.head()

Unnamed: 0                                               text  toxic  \
0           0  Explanation Why the edits made under my userna...      0   
1           1  D'aww He matches this background colour I'm se...      0   
2           2  Hey man I'm really not trying to edit war It's...      0   
3           3  More I can't make any real suggestions on impr...      0   
4           4  You sir are my hero Any chance you remember wh...      0   

                                           lemm_text  
0  Explanation Why the edits made under my userna...  
1  D'aww He match this background colour I 'm see...  
2  Hey man I 'm really not trying to edit war It ...  
3  More I ca n't make any real suggestion on impr...  
4  You sir are my hero Any chance you remember wh...

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>
Так же можешь попробовать использовать spacy:
    
```python
import spacy
nlp = spacy.load('en_core_web_sm')

def lemmatize_spacy(text):
    return " ".join([token.lemma_ for token in nlp(text)])

data['lemm_spacy'] = data['text'].progress_apply(lemmatize_spacy)     
```

Вот так можно следить за выполнением функции (как раз выше его использовал)

```python
    
from tqdm import tqdm
tqdm.pandas()
data['text'] = data['text'].progress_apply(lemmatize_text) 
```
Где *lemmatize_text* - функция

    
P.S Кстати можно ускорить spacy ( раз в 5 +-)
    
```python
new_corpus = []

for doc in tqdm(nlp.pipe(data['text'], batch_size=64, n_process=-1, disable=["parser", "ner"]), total=len(data['text'])):
    word_list = [tok.lemma_ for tok in doc]
    new_corpus.append(' '.join(word_list))
    
data['lemm_spacy_new'] = new_corpus   
    
    
```
    
Для больших объемов текста ```SpaCy``` рекомендует использовать ```nlp.pipe```, который может работать в пакетах ```batch_size```(Допустим ```batch_size=64```, тогда все наши данные делятся на 64 пакета (как прям в фолдах, только там мы пишем число на сколько поделить наши данные (3/5/10 частей),а тут пишем, сколько данных отдать в обработку)  и имеет встроенную поддержку многопроцессорной обработки ```n_process``` (аналог ```n_jobs``` для GridSearchCV).

Кроме того, нужно убедиться, что мы отключили все элементы конвейера, которые мы не планируем использовать, поскольку они просто потратят время на обработку. Если мы выполняем только лемматизацию, то необходимо передать ```disable=["parser", "ner"]``` к ```nlp.pipe```.
    
    
P.S.S Если запустишь ```nlp.pipe```, то первое время нужно подождать прежде, чем начнется обработка 
</div>

In [8]:
state = np.random.RandomState(12345) #фиксируем значение для упрощения кода в дальнейшем

In [9]:
target = data['toxic']
features = data.drop(['toxic'], axis=1)

features_train, features_test, target_train, target_test = train_test_split(features, 
                                                                              target, 
                                                                              test_size=0.2, 
                                                                              random_state=state)

In [10]:
nltk.download('stopwords')
stopwords = set(nltk_stopwords.words('english'))


[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [11]:
count_tf_idf = TfidfVectorizer(stop_words=stopwords)

features_train = count_tf_idf.fit_transform(features_train['lemm_text'].values.astype('U'))
features_test = count_tf_idf.transform(features_test['lemm_text'].values.astype('U'))

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Перевод в unicode актуально для русского языка, который был в тренажере, в данной задаче это делать нет необходимости

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Отлично, молодец, верно используешь Tfidf.
    
    
Совет: Внутри кросс-валидации происходит разбиение выборки на train и valid. Однако, в таком случае векторизатор обучен на всей выборке(train), а это не совсем корректно. Чтобы избежать это можно воспользоваться Pipeline:
    
```python
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(stop_words=stopwords)),
    ('logreg', LogisticRegression(random_state=42)),
])
parameters = {
    'tfidf__max_df': (0.25, 0.5, 0.75),
    'tfidf__ngram_range': [(1, 1), (1, 2)],
    'logreg__C': [1,2,6]
}

grid_search_tune = RandomizedSearchCV(pipeline, parameters, cv=3, n_jobs=-1, scoring='f1', verbose=3)
grid_search_tune.fit(train_features, train_targets)
  
    
```
    
Это просто каркас, можешь сам выбрать какие параметры использовать для подбора:) 
    
+  https://runebook.dev/ru/docs/scikit_learn/modules/generated/sklearn.model_selection.halvinggridsearchcv - тут про HalvingGridSearchCV
    
+  https://www.rupython.com/python-sklearn-pipeline-pipeline-28301.html - про pipeline
+  https://towardsdatascience.com/how-to-use-sklearn-pipelines-for-ridiculously-neat-code-a61ab66ca90d

</div>

In [12]:
print(features_train.shape)
print(features_test.shape)

(127433, 143414)
(31859, 143414)


## Обучение

###  LogisticRegression

In [13]:
logistic_model = LogisticRegression()
param_grid = {'solver':['newton-cg','liblinear'],
             'class_weight':['balanced'],
             'C' : [10]}
grid_search = GridSearchCV(logistic_model, param_grid, scoring='f1',cv=3)
grid_search.fit(features_train, target_train)
LR_best_params = grid_search.best_params_
print('F1 ', grid_search.best_score_)
LR_best_params

F1  0.7598132729892986


{'C': 10, 'class_weight': 'balanced', 'solver': 'newton-cg'}

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Хорошо, но почему C:10? Все же стояло по перебирать параметры, например: 
    
+  ```class_weight: [None, 'balanced']``` - да, у нас есть дисбаланс, но вдруг модель и без изменения весов получит лучше результат
+  ```C: range(1, 11, 2)``` - от 1 до 11 с шагом 2 это параметр регуляризации (должен помочь)

</div>

### DecisionTreeClassifier

In [14]:
param_grid = {
    'max_depth': [100], 
    'class_weight':['balanced']
}
tree_model = DecisionTreeClassifier(random_state=state)
grid_search = GridSearchCV(estimator=tree_model, param_grid=param_grid,cv=3,scoring='f1')
grid_search.fit(features_train, target_train)
best_tree = grid_search.best_params_    
print('F1 ', grid_search.best_score_)
best_tree

F1  0.6377649759281363


{'class_weight': 'balanced', 'max_depth': 100}

### LGBMClassifier

In [15]:
param_grid = {'max_depth' : [10],
            'n_estimators' : [100]
}
lgbm_model = LGBMClassifier(random_state=state)
grid_search = GridSearchCV(estimator=lgbm_model, param_grid=param_grid,cv=3,scoring='f1')
grid_search.fit(features_train, target_train)
best_lgbm = grid_search.best_params_

print('F1:', grid_search.best_score_)
print('Лучшие параметры ', best_lgbm)

F1: 0.6827763500163079
Лучшие параметры  {'max_depth': 10, 'n_estimators': 100}


### CatBoostClassifier - не вытягивает ядро тренажера, каждый раз при запуске пишет, "Ядро умерло"

In [16]:
#train_featurescat =  features_train.toarray() 
#valid_featurescat =  features_valid.toarray() 
#test_featurescat = features_test.toarray() 
#cat_model = CatBoostClassifier(eval_metric="F1", 
#                                   iterations=100, 
#                                   max_depth=10)
#cat_model.fit(train_featurescat, target_train)
#cat_prediction = cat_model.predict(train_featurescat)
#f1 = f1_score(target_train, cat_prediction)
#print('F1:', f1_score)

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Так, а зачем переводить в array?

Для Catboost будет лучше на самом деле воспользовать внутренним кодированием:
    
    
+  https://colab.research.google.com/drive/1bQOXIKpeyGZgda2qQqDijE2SIbSGX7-Z?usp=sharing   
    
Только вначале необходимо настроить `GPU` в `Google Colab`, так получится обработать все данные +- за 5-20 минут 
    
    
+  `Среда выполнения` -> `сменить среду выполнения`    
+  Теперь выбираем `GPU` и `сохранить`  
    
P.S Ну или попробовать на CPU
    
    
 **не** стоит проводить тестирование на обучающей выборке, потому что мы получим позитивный результат, так как модель уже видела эти данные 
</div>

## Тестирование

In [17]:
model = LogisticRegression()
model.set_params(**LR_best_params)
model.fit(features_train, target_train)
prediction = model.predict(features_test)
f1 = f1_score(target_test, prediction)
print('F1', f1)
print('Матрица ошибок')
print(confusion_matrix(target_test, prediction))

F1 0.7671942446043165
Матрица ошибок
[[27575  1054]
 [  564  2666]]


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 

Верно проверил тестирование
    
</div>

Показатели данной модели удовлетворяют нашему условию и на тестовой выборке, значит мы однозначно можем ее рекомендовать.

## Выводы

In [20]:
df = [  [0.76],
        [0.64],
        [0.68]]
model = ["LogisticRegression", "DecisionTree", "LGBMRegressor"]
pd.DataFrame(data=df, index=model, columns=["F1"])

F1
LogisticRegression  0.76
DecisionTree        0.64
LGBMRegressor       0.68

В ходе работы мы подготовили данные к обучению. Разобрались, что в данных присутствует дисбаланс. Получили список лемматизированных слов и очистили лишние символы. Закодировали в юникоде, для качественного обучения

Разделили на тренировочную и тестовую выборку

Мы обучили несколько моделей и лучшей из них стала LogisticRegression с гиперпараметрами ('C': 10, 'class_weight': 'balanced', 'solver': 'newton-cg'), которая имеет меньше ложно отрицательных предсказаний и F1 = 0.7598 на тренировочной и 0.767 на тестовой.

Можно сказать, что количество ложно положительных предсказаний у LogisticRegression не является проблемой, так как эти комментарии могут быть идентифицированы и отклонены после модерации. Итак, для наших бизнес-целей LogisticRegression является лучшей моделью.

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 

Для работы с текстами используют и другие подходы. Например, сейчас активно используются RNN (LSTM) и трансформеры (BERT и другие с улицы Сезам, например, ELMO). НО! Они не являются панацеей, не всегда они нужны, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут справляться. \
BERT тяжелый, существует много его вариаций для разных задач, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множество реализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 
    
</div>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Данные загружены и подготовлены
- [ ]  Модели обучены
- [ ]  Значение метрики *F1* не меньше 0.75
- [ ]  Выводы написаны

<font color='blue'><b>Итоговый комментарий ревьюера</b></font>
<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>Софья, спасибо за хороший проект!!! Я готов принять работу, но хочу убедиться, что тебе все понятно.<br>
Если есть какие либо вопросы я с удовольствием на них отвечу:)</div>


<div class="alert alert-success">
<b>Комментарий ревьюера V2✔️:</b> Удачи в следующих проектах!!!
</div>